database/database.go

```diff
@@ -1,127 +1,585 @@
+	Upvotes   uint32
+	Match     float32
 }
 
-// Quote struct
-type Quote struct {
-	ID      int
-	Teacher Teacher
-	Text    string
+// UnverifiedQuoteT stores one unverified quote
+// QuoteID    the unique identificator of the unverified quote
+// teacher 	   the name of the teacher
+// Context	   the context of the quote
+// text        the text of the quote itself
+// unixtime    optional
+// iphash	   optional
+//
+type UnverifiedQuoteT struct {
+	QuoteID  uint32
+	Teacher  string
+	Context  string
+	Text     string
+	Unixtime uint64
+	IPHash   uint64
 }
 
-var database *sql.DB
+// TeacherT stores one teacher
+// TeacherID the unique identifier of the teacher
+// name       the teacher's name
+// title      the teacher's title
+// note       optional notes, e.g. subjects
+type TeacherT struct {
+	TeacherID uint32
+	Name      string
+	Title     string
+	Note      string
+}
+
+type wordsMapT struct {
+	totalOccurences uint32
+	occurenceSlice  []occurenceSliceT
+}
+
+type occurenceSliceT struct {
+	enumid int32
+	count  uint32
+}
+
+/* -------------------------------------------------------------------------- */
+/*                          GLOBAL PACKAGE VARIABLES                          */
+/* -------------------------------------------------------------------------- */
+
+// Handle to the PostgreSQL database, used as long time storage
+var postgresDatabase *sql.DB
 
-// SetupDatabase is a function to setup the database
-func SetupDatabase() {
+// Created from PostgreSQL database at (re)start
+// localDatabase is a cache of the postgreSQL database to speed up read operations
+//
+// unverified quotes will not be cached in the local database, because read operations
+// will only be performed by the operator and thus be very rare
+//
+// important: the unverifiedQuote table will not be mirrored
+//
+// important: the index of a quote in quoteSlice is called its enumid
+// which is used to quickly identify a quote with the wordsMap
+var localDatabase struct {
+	quoteSlice   []QuoteT
+	teacherSlice []TeacherT
+	wordsMap     map[string]wordsMapT
+	mux          Mutex
+}
+
+/* -------------------------------------------------------------------------- */
+/*                         EXPORTED GENERAL FUNCTIONS                         */
+/* -------------------------------------------------------------------------- */
+
+// Setup initializes the database backend
+// Initialize postgres database
+// Create localDatabase from postgresDatabase
+//
+// Must be called only once at startup before any of the other database functions
+func Setup() error {
 	var err error
 
-	database, err = sql.Open("postgres", "user=postgres password=1234 dbname=quote_gallery sslmode=disable")
+	// Open PostgreSQL database
+	postgresDatabase, err = sql.Open(
+		"postgres",
+		`user=postgres 
+		password=1234 
+		dbname=quote_gallery 
+		sslmode=disable`)
 	if err != nil {
-		log.Fatal("Cannot open Database: ", err)
+		return errors.New("At Setup: " + err.Error())
```

Comment:
(goes for the whole file): try to use more descriptive error message to help the person running the app.
This is just an example below - always state which operation failed.
```suggestion
		return errors.New("Connecting to Database failed in Setup(): " + err.Error())
```

Reply:
I'll do this